### PR TITLE
preprocess: Allow preprocessing any file extensions

### PR DESCRIPTION
**Pull request by xclaesse**

### CODE REVIEW

File: mesonbuild/backend/ninjabackend.py [hunk 1]
```diff
@@ -830,6 +830,10 @@ def generate_target(self, target):
             self.generate_swift_target(target)
             return
 
+        # CompileTarget compiles all its source and does not do a final link.
```
Comment (tristan957) [inline mesonbuild/backend/ninjabackend.py:833]: sources?

File: mesonbuild/backend/ninjabackend.py [hunk 2]
```diff
@@ -830,6 +830,10 @@ def generate_target(self, target):
...
+        # This is for example a preprocessor.
```
Comment (tristan957) [inline mesonbuild/backend/ninjabackend.py:834]: This is, for example, a preprocessor

**Comment by willwray** [comment]
Just a note to say thtat it looks like this PR addresses an issue I'd hit with preprocess, thanks.

I'd previously used a custom_target in which input and output files both had `.hpp` extension.
In porting to cc.preprocess, I had to change the input file extension, or it wasn't preprocessed.
In doing so, I lost the convenient automatic syntax highlighting of a recognized source file extension.

File: mesonbuild/backend/ninjabackend.py (hunk 1 above)
Comment (xclaesse) [reply mesonbuild/backend/ninjabackend.py]: done

File: mesonbuild/backend/ninjabackend.py (hunk 2 above)
Comment (xclaesse) [reply mesonbuild/backend/ninjabackend.py]: done

**Comment by xclaesse** [comment]
I've hit this issue when using cc.preprocess in fontconfig: https://gitlab.freedesktop.org/fontconfig/fontconfig/-/merge_requests/275. Would be great to get it reviewed :)

**Review by tristan957 — APPROVED** [review]

**Pull request merged**